nexus/src/db/db-macros/src/lookup.rs

```diff
@@ -0,0 +1,649 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Procedure macro for generating lookup structures and related functions
+//!
+//! See nexus/src/db/lookup.rs.
+
+use proc_macro2::TokenStream;
+use quote::{format_ident, quote};
+
+//
+// INPUT (arguments to the macro)
+//
+
+/// Arguments for [`lookup_resource!`]
+// NOTE: this is only "pub" for the `cargo doc` link on [`lookup_resource!`].
+#[derive(serde::Deserialize)]
+pub struct Input {
+    /// Name of the resource (PascalCase)
+    name: String,
+    /// ordered list of resources that are ancestors of this resource, starting
+    /// with the top of the hierarchy
+    /// (e.g., for an Instance, this would be `[ "Organization", "Project" ]`
+    ancestors: Vec<String>,
+    /// unordered list of resources that are direct children of this resource
+    /// (e.g., for a Project, these would include "Instance" and "Disk")
+    children: Vec<String>,
+    /// describes how the authz object for a resource is constructed from its
+    /// parent's authz object
+    authz_kind: AuthzKind,
+}
+
+/// Describes how the authz object for a resource is constructed from its
+/// parent's authz object
+///
+/// By "authz object", we mean the objects in nexus/src/authz/api_resources.rs.
+///
+/// This ought to be made more uniform with more typed authz objects, but that's
+/// not the way they work today.
+#[derive(serde::Deserialize)]
+enum AuthzKind {
+    /// The authz object is constructed using
+    /// `authz_parent.child_generic(ResourceType, Uuid, LookupType)`
+    Generic,
+
+    /// The authz object is constructed using
+    /// `authz_parent.$resource_type(Uuid, LookupType)`.
+    Typed,
+}
+
+//
+// MACRO STATE
+//
+
+/// Configuration for [`lookup_resource`] and its helper functions
+///
+/// This is all computable from [`Input`].  This precomputes a bunch of useful
+/// identifiers and token streams, which makes the generator functions a lot
+/// easier to grok.
+pub struct Config {
+    // The resource itself that we're generating
+    /// Basic information about the resource we're generating
+    resource: Resource,
+
+    /// See [`AuthzKind`]
+    authz_kind: AuthzKind,
+
+    // The path to the resource
+    /// list of type names for this resource and its parents
+    /// (e.g., [`Organization`, `Project`])
+    path_types: Vec<syn::Ident>,
+
+    /// list of identifiers used for the authz objects for this resource and its
+    /// parents, in the same order as `authz_path_types`
+    /// (e.g., [`authz_organization`, `authz_project`])
+    path_authz_names: Vec<syn::Ident>,
+
+    // Child resources
+    /// list of names of child resources (PascalCase, raw input to the macro)
```

Comment:
Again, I don't think the "PascalCase" bit is the piece that matters - these are expected to be the names of types / structures which exist in specific modules, and need specific functions.